### PR TITLE
CopyProp: Propagate moves of constants

### PR DESCRIPTION
**Pull request by clubby789**
Yet another try of #120650 - this time, keep track of SSA locals with a constant value and propagate it. Then, re-run `SimplifyConstCondition` to take advantage of it (simply reordering the pass regressed some tests).

r? @ghost

### CODE REVIEW

**Comment by clubby789** [comment]
Removed the pass ordering part. If perf looks happier, perhaps we can use this to simplify #120650 - rather than needing to `codegen_operand`, we can just check if our discrim is a const scalar.

@bors try @rust-timer queue

**Comment by clubby789** [comment]
Small net improvements if I run perf locally, so I think just noise

**Comment by clubby789** [comment]
It looks like the previous slight improvements to MIR in tests were also done by #120594. That said, this seems to be basically perf-neutral and a bit more general so may be worth having.

File: compiler/rustc_mir_transform/src/copy_prop.rs [hunk 1]
```diff
@@ -75,31 +79,44 @@ fn fully_moved_locals(ssa: &SsaLocals, body: &Body<'_>) -> BitSet<Local> {
     let mut fully_moved = BitSet::new_filled(body.local_decls.len());
 
     for (_, rvalue, _) in ssa.assignments(body) {
-        let (Rvalue::Use(Operand::Copy(place) | Operand::Move(place))
-        | Rvalue::CopyForDeref(place)) = rvalue
-        else {
+        let (Rvalue::Use(Operand::Copy(place)) | Rvalue::CopyForDeref(place)) = rvalue else {
```
Comment (clubby789) [inline compiler/rustc_mir_transform/src/copy_prop.rs:82]: Given that the `Operand::Move` case does nothing (due to the check below), we can remove that here and remove the duplicated check below

**Comment by compiler-errors** [comment]
r? cjgillot

**Comment by clubby789** [comment]
First lets see if this still has any perf impact:

@bors try @rust-timer queue

I suspect not - if this comes up neutral I'll probably close

**Pull request closed without merging**